acareeg/eegip.py

```diff
@@ -171,53 +175,122 @@ def process_events_london_resting_state(raw, age, tmax=1):
-        annot_sample = (annot_sample * freq).astype(int)
+        # Keep only trials that are at least 10 s long.
+        # All trials are defined by Rst0-Rst1 pair of events.
+        annot_sample = np.concatenate([np.arange(start*freq, (stop-tmax)*freq, tmax*freq)
+                                       for start, stop in zip(rst0, rst1)
+                                       if stop - start > 10.0]).astype(int)
         annot_id = [1] * len(annot_sample)
 
     elif age == 12:
-        annots = [a for a in raw.annotations if a["description"] in ["eeg1", "eeg2", "eeg3"]]
-        if len(annots) == 0:
+        """
+         For London 12m, resting-state data are marked only by Rst0 markers at the begining of
+         every trials. Trials and consecutives, so we can use the following Rst0 maker as being the 
+         end of the current trials. The last trials is ended with the last occurence of a SWIR or SWEN
+         event. 
+         
+         Trials shorter than 10s are rejected. Trials longer than 50s are truncated at 50s.
+         
+         Events eeg1, eeg2 and eeg3, when present, are at the same time as the Rst0 events. They 
+         are used to set the type of event (type of video that was presented). If no eegx event
+         is available for a Rst0, as type eeg4 is set, which indicate a resting-state event using
+         an unknown video.
+        """
+
+        onsets = [annot["onset"] for annot in raw.annotations if annot["description"] == "Rst0"]
+        if len(onsets) == 0:
             return None
 
-        if raw.annotations[-1]["onset"] > annots[-1]["onset"]:
-            end = np.min([raw.annotations[-1]["onset"], annots[-1]["onset"] + 50.])
-            annots.append(OrderedDict((("onset", end), ("duration", 0),
-                                       ("description", "end"), ('orig_time', None))))
-
-        for annot, next_annot in zip(annots[:-1], annots[1:]):
-            annot_sample.append(np.arange(int(annot["onset"] * freq),
-                                          int(next_annot["onset"] * freq),
+        final_swir_swen_onset = [annot["onset"]
+                                 for annot in raw.annotations
+                                 if annot["description"] in ["SWEN", "SWIR"]][-1]
+        durations = list(np.diff(onsets)) + [final_swir_swen_onset - onsets[-1]]
+
+        # Truncate longer trials at 50s
+        durations = [min(d, 50) for d in durations]
+
+        # Reject any trial that last less than 10s
+        durations, onsets = zip(*[(d, onset) for d, onset in zip(durations, onsets) if d >= 10])
+
+        # Find eeg1, eeg2, eeg3 events clost (less than 5s appart) for Rst0 and attribute this event
+        # type. Else, attribute eeg4 (unpecified video type).
+        eegx_annots = [(a["description"], a["onset"])
+                       for a in raw.annotations
+                       if a["description"] in ["eeg1", "eeg2", "eeg3"]]
+        if len(eegx_annots):
```

Comment:
if there are eegx_annots values, then eegx_desc, eegx_onsets are equal to these eegx_annots values? otherwise eegx_desc and eegx_onsets stay empty.

Reply:
Yes. These two list are used below to associate the eegx description to the Rst0 markers.